### DISCUSSION TITLE
Help with using BA-DUA

### DISCUSSION THREAD

**Discussion by Alex-128** [saeg/ba-dua#37] [Q&A] [answered]
Good day

I am a Computer Science student in South Africa, attending Stellenbosch University, and for my year-project, I'm focusing on spectrum-based fault localization using data-flow coverage. In the search for data-flow coverage tools, BA-DUA immediately came to my attention due to the number of articles that mention it (as well as the fact that it's open-source!). I've downloaded this source code and tried to run it for an example Java class following the instructions in the README, but I can't get it to work. When I run the 'instrument' command, it prints out, "0 classes instrumented in 0 seconds", whether I give it a source folder with Java classes (and compiled classes) or the name of a compiled class. I then tried to run the 'report' command, but it prints out, "Failed: Invalid execution data file.", probably due to the file I give as -input being uninstrumented.

I would really like to use your tool for my project, but I'm struggling to understand where I'm going wrong. If you have any help or advice to offer, I'd really appreciate it.

Kind regards
Alexandra van der Spuy

**Comment by andrioli** [comment]
Nice to hear that you are planning to use BA-DUA.

Can you provide the version of Java and BA-DUA are you using? currently BA-DUA 0.7.0 supports classes compiled up to Java 19. Yesterday I released BA-DUA 0.8.0 that supports classes compiled up to Java 20.

Please, also provide a transcript of the steps that you tried to execute. Normally the `0 classes instrumented in 0 seconds` is due to the command not able to find any compatible compiled classes. Note that the classes should be compiled first with `javac` or using any helping building tool like Ant, Maven, etc.

The `report` can only read files that were generated by the instrumented program execution. So we need to instrument and execute the classes first.

Regards

Roberto

**Reply by Alex-128** [reply]
Thank you for the quick reply, I really appreciate it!

I downloaded the latest BA-DUA on GitHub, which seems to be 0.7.0, and I'm using Java 11. However, I don't know what I was doing wrong before, because I just compiled my Maven test project and passed the target/classes/com/example (which contains the compiled class) to `instrument`, and it said `1 classes instrumented in 0 seconds`. So that part is working!

High on this success, I tried the `report` phase, but unfortunately my luck ran out :( It still says `Failed: Invalid execution data file.`
I'm not completely sure about the input I'm supposed to give for this phase. The output says `-classes` means the directory containing the instrumented classes, so I pass it, for example, the directory `instrumented`, which was my `-dest` during the `instrument` phase. What should be the `-input` argument, though? I've passed it an instrumented class, an uninstrumented class, the original `.java` file, the test class, and the test `.java` file, but all of them give the same output.

Thank you once more.

Kind regards
Alexandra

**Reply by andrioli** [reply, marked as answer]
> I downloaded the latest BA-DUA on GitHub, which seems to be 0.7.0, and I'm using Java 11. However, I don't know what I was doing wrong before, because I just compiled my Maven test project and passed the target/classes/com/example (which contains the compiled class) to `instrument`, and it said `1 classes instrumented in 0 seconds`. So that part is working!

Perfect, it seems to be correct (if `-src` folder contains only one class). Also note that the `instrument` program traverses the directories recursively, providing only the `target/classes` should be enough.  It will use the same folder structure for the destination directory, so every instrumented class will live in the correct place.

> The output says `-classes` means the directory containing the instrumented classes, so I pass it, for example, the directory `instrumented`, which was my `-dest` during the `instrument` phase.

The command is not well explained so, actually for `-classes` you should provide the path with the original classes (the non-instrumented ones). You should use same as `-src` and not `-dest`.

> What should be the `-input` argument, though?

You need yo execute you program first because `-input` is the binary file generated by BA-DUA after instrumented classes execution.  You can execute for instance, the unit tests or a `main` method. The only important thing here is that  the instrumented classes should be executed by he JVM. In this case, after execution,  a new file will be created in the current working directory.  By default the file name is `coverage.ser`, but you can use whatever name you want by setting a Java system property `output.file`.

---

A PR clarifying the use of the CLI options will be very welcome 😄. Here you can find the files the the CLI options descriptions:
* https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-cli/src/main/java/br/usp/each/saeg/badua/cli/InstrumentOptions.java
* https://github.com/saeg/ba-dua/blob/ba-dua-0.8.0/ba-dua-cli/src/main/java/br/usp/each/saeg/badua/cli/ReportOptions.java

We also have this very old [repository](https://github.com/saeg/ba-dua-example) with examples of BA-DUA usage. There you can find an [example using the offline instrumentation](https://github.com/saeg/ba-dua-example/blob/master/build.sh) and also [Maven](https://github.com/saeg/ba-dua-example/blob/master/pom.xml) and [Ant](https://github.com/saeg/ba-dua-example/blob/master/build.xml) examples, but these two using [Java agent](https://docs.oracle.com/javase/7/docs/api/java/lang/instrument/package-summary.html) instead.

I hope this examples are still valid today and these instructions will help you

Best Regards

Roberto

**Reply by Alex-128** [reply]
Hi, thanks for the extra information! Sorry for the radio silence, unfortunately the university assignments were piling up and I didn't have as much time to spend on this as I wanted to. I had a look at these links and they're very helpful, so thank you once more. However, when I followed your instructions and executed the instrumented files, it gave me the exception
```
Exception in thread "main" java.lang.NoClassDefFoundError: br/usp/each/saeg/badua/agent/rt/internal_6cf338b/Offline
```
I was initially worried that it was an issue on my side, but my classes have main methods and the non-instrumented programs run perfectly fine. Is there something I can do to fix this issue?
Sorry for sending this just before the long weekend, feel free to ignore it until it suits you!

Kind regards
Alexandra

**Reply by andrioli** [reply]
> Hi, thanks for the extra information! Sorry for the radio silence, unfortunately the university assignments were piling up and I didn't have as much time to spend on this as I wanted to.

Don't worry, I know exactly how it is.

> However, when I followed your instructions and executed the instrumented files, it gave me the exception
> ```
> Exception in thread "main" java.lang.NoClassDefFoundError: br/usp/each/saeg/badua/agent/rt/internal_6cf338b/Offline
> ```

This is because the instrumented classes have a direct dependency of the BA-DUA runtime classes.  So you need to include BA-DUA jar to the classpath (the option `-cp` or `-classpath` when running  `java` command). Just include the JAR `ba-dua-cli-all` or `ba-dua-agent-rt-all` to your Java classpath.

**Comment by Alex-128** [comment]
Thanks again! I thought that was the problem and tried to include the JAR before sending you the previous message, but I got a bit confused about how to include it in the classpath so it was still giving errors. But I read up on the classpath thing and it worked when I tried it now! The report also worked :) 

As for my project, this tool definitely seems like a good option, because I want to work with the raw data-flow coverage instead of something colour-coded. So congratulations, I guess! Thanks for all the help and goodbye for now, but I'll definitely return if I have more questions, if that's all right?

**Comment by andrioli** [comment]
I'm glad it worked.

Current version of BA-DUA can export a XML or a CSV reports (the contents from stdout). 

I'm also working on a Java API to read the binary file that contains the coverage info but this is not ready yet.

I'll close this discussion as resolved. If you have any other doubts, fell free to open a new discussion.
